tensorboard/program.py

```diff
@@ -341,79 +342,128 @@ def __init__(self, msg):
     self.msg = msg
 
 
+class TensorBoardFailedToBindError(TensorBoardServerException):
+  """Error raised when failing to bind to a port.
+
+  This should be raised when it is expected that binding to another
+  similar port would succeed---for instance, if the underlying error is
+  that the address is in use, not that the address family is
+  unsupported. This is used as a signal to indicate that automatic port
+  searching should continue rather than aborting.
+  """
+  pass
+
+
+def with_port_scanning(cls):
```

Comment:
The class decorator is nifty, but it seems to have some functional disadvantages versus just having `with_port_scanning` return `__init__` rather than a wrapper class, and then define `create_port_scanning_werkzeug_server = with_port_scanning(WerkzeugServer)` and making that the default value for `TensorBoard.server_class` (possibly renamed to server_factory).

- it's more conceptually complicated since class decorators that implement functional behavior themselves are somewhat exotic
- it breaks super()
- it requires re-implementing pass-through versions of server_forever() and get_url() that we'd otherwise get for free
- it makes the original WerkzeugServer inaccessible to e.g. test code or other uses

If you're particularly enthusiastic about this approach it's ok, but personally it seems better to keep it simple?

Reply:
> return `__init__` rather than a wrapper class

Good point. That should work. Let me try it.

Reply:
Done. Kept the name `server_class` because that would technically be a
breaking change, and the functionality didn’t actually change (you could
have passed a factory the whole time).